### ISSUE TITLE
Resharding on restart failed with out of space

### ISSUE THREAD

**Issue by enaydanov** [scylladb/scylla#10496]
## Installation details

Kernel Version: 5.13.0-1021-aws
Scylla version (or git commit hash): `2022.1~rc3-20220406.5cc3b678c` with build-id `48dfae0735cd8efc4ae2f5c777beaee2a1e89f4a`
Cluster size: 4 nodes (i3.4xlarge)

Scylla Nodes used in this run:
- longevity-tls-1tb-7d-2022-1-db-node-10419cea-9 (34.244.12.201 | 10.0.3.117) (shards: 14)
- longevity-tls-1tb-7d-2022-1-db-node-10419cea-8 (54.246.46.66 | 10.0.3.118) (shards: 16)
- longevity-tls-1tb-7d-2022-1-db-node-10419cea-7 (54.195.187.103 | 10.0.3.122) (shards: 14)
- longevity-tls-1tb-7d-2022-1-db-node-10419cea-6 (34.245.131.65 | 10.0.2.140) (shards: 16)
- longevity-tls-1tb-7d-2022-1-db-node-10419cea-5 (3.250.105.236 | 10.0.2.227) (shards: 14)
- longevity-tls-1tb-7d-2022-1-db-node-10419cea-4 (54.195.181.225 | 10.0.0.60) (shards: 14)
- longevity-tls-1tb-7d-2022-1-db-node-10419cea-3 (3.250.194.99 | 10.0.1.41) (shards: 14)
- longevity-tls-1tb-7d-2022-1-db-node-10419cea-2 (63.33.212.26 | 10.0.1.52) (shards: 14)
- longevity-tls-1tb-7d-2022-1-db-node-10419cea-13 (54.216.102.112 | 10.0.3.149) (shards: 14)
- longevity-tls-1tb-7d-2022-1-db-node-10419cea-12 (34.254.205.161 | 10.0.2.156) (shards: 14)
- longevity-tls-1tb-7d-2022-1-db-node-10419cea-11 (34.245.137.240 | 10.0.1.231) (shards: 14)
- longevity-tls-1tb-7d-2022-1-db-node-10419cea-10 (18.203.224.95 | 10.0.1.175) (shards: 14)
- longevity-tls-1tb-7d-2022-1-db-node-10419cea-1 (54.170.174.72 | 10.0.2.128) (shards: 14)

OS / Image: `ami-071c70d20f0fdbb2c` (aws: eu-west-1)

Test: `longevity-1tb-7days-test`
Test id: `10419cea-c7fd-4674-baef-09e313deea47`
Test name: `enterprise-2022.1/longevity/longevity-1tb-7days-test`
Test config file(s):

- [longevity-1TB-5days-authorization-and-tls-ssl.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/26669988977199aa81816c32f017be68c3fc0dea/test-cases/longevity/longevity-1TB-5days-authorization-and-tls-ssl.yaml)

## Issue description

RestartWithResharding nemesis on node longevity-tls-1tb-7d-2022-1-db-node-10419cea-6 failed with "No space left" error:

```
...
May 02 16:42:50 longevity-tls-1tb-7d-2022-1-db-node-10419cea-6 scylla[41896]:  [shard 0] database - Resharding 971GB for keyspace1.standard1
...
May 02 17:02:05 longevity-tls-1tb-7d-2022-1-db-node-10419cea-6 scylla[41896]:  [shard 8] storage_service - Shutting down communications due to I/O errors until operator intervention: Disk error: std::system_error (error system:28, No space left on device)
...
```

![image](https://user-images.githubusercontent.com/63434/167090144-be6a398f-32e8-4525-bfcb-c037fe88d3b3.png)

- Restore Monitor Stack command: `$ hydra investigate show-monitor 10419cea-c7fd-4674-baef-09e313deea47`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=10419cea-c7fd-4674-baef-09e313deea47)
- Show all stored logs command: `$ hydra investigate show-logs 10419cea-c7fd-4674-baef-09e313deea47`

## Logs:
- **db-cluster-10419cea.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/10419cea-c7fd-4674-baef-09e313deea47/20220502_174300/db-cluster-10419cea.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/10419cea-c7fd-4674-baef-09e313deea47/20220502_174300/db-cluster-10419cea.tar.gz)
- **monitor-set-10419cea.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/10419cea-c7fd-4674-baef-09e313deea47/20220502_174300/monitor-set-10419cea.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/10419cea-c7fd-4674-baef-09e313deea47/20220502_174300/monitor-set-10419cea.tar.gz)
- **loader-set-10419cea.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/10419cea-c7fd-4674-baef-09e313deea47/20220502_174300/loader-set-10419cea.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/10419cea-c7fd-4674-baef-09e313deea47/20220502_174300/loader-set-10419cea.tar.gz)
- **sct-runner-10419cea.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/10419cea-c7fd-4674-baef-09e313deea47/20220502_174300/sct-runner-10419cea.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/10419cea-c7fd-4674-baef-09e313deea47/20220502_174300/sct-runner-10419cea.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/enterprise-2022.1/job/longevity/job/longevity-1tb-7days-test/4/)

**Comment by roydahan** [comment]
@slivne this is a bit weird, we use LCS in this test with apx. 1TB per node.

**Comment by slivne** [comment]
![Screenshot from 2022-05-11 13-57-14](https://user-images.githubusercontent.com/3465480/167833983-6cccec5a-96b7-40c1-91c2-7d600ef831c3.png)

Prior to reshape node was at ~60% - so we had 40% left.

In resharding we may need to rewrite all the data,

@bhalevy / @raphaelsc how does is resharding work - do we delete tables after they have been read and resharded or do we wait till the end and only then delete ?

**Comment by slivne** [comment]
@roydahan is this a regression in your view

**Comment by bhalevy** [comment]
> @bhalevy / @raphaelsc how does is resharding work - do we delete tables after they have been read and resharded or do we wait till the end and only then delete ?

From https://github.com/scylladb/scylla/blob/6b34ba3a4fff4f0fc0ce9b88166d2482db2dc076/replica/distributed_loader.cc#L204-L206

and https://github.com/scylladb/scylla/blob/6b34ba3a4fff4f0fc0ce9b88166d2482db2dc076/sstables/sstable_directory.cc#L405-L417

I understand that each resharding compaction job runs on a maximum of `max_threshold` (32) sstables and will delete them when finished. BUT, we run all jobs in parallel, so we generate a lot of rewritten sstables at once.

We need to serialize the jobs or at least limit their parallelism.

**Comment by bhalevy** [comment]
Hmm https://github.com/scylladb/scylla/blob/6b34ba3a4fff4f0fc0ce9b88166d2482db2dc076/compaction/compaction_manager.cc#L265

**Comment by bhalevy** [comment]
Maybe there's a shared ptr leak so the files are still opened and the storage is not released.

**Comment by raphaelsc** [comment]
I did some analysis last week, and there seems no regression in reshard. I'll check if there's possible a regression elsewhere.

**Comment by enaydanov** [comment]
Used bytes for 4.6.3 (5 days of RestartWithResharding-only nemesis):

![image](https://user-images.githubusercontent.com/63434/172040262-f5e01f59-1202-4e62-b858-d02694808fce.png)

**Comment by enaydanov** [comment]
4.6.3 repro run with same nemesis seed:

![image](https://user-images.githubusercontent.com/63434/172041122-a1fb7ecd-e892-4be2-8995-d2a1efb64c5a.png)

On this picture we can see significant growth of used space after CorrupThenScrub nemesis.

**Comment by enaydanov** [comment]
The used bytes metrics from the original test run looks like following:

![image](https://user-images.githubusercontent.com/63434/172042425-5758f416-587f-4119-93e2-95f3bec46648.png)

Same growth after CorruptThenScrub and another growth for orange one after StartStopScrubCompaction.

It doesn't looks like a regression and not related to resharding from this POV. More about what happen after the scrub.

As I can see from the description of "nodetool scrub" operation it creates a snapshot. Should we do some clean up after?

**Comment by roydahan** [comment]
@enaydanov Yes, we should clean any place we're taking a snapshot when we're "done with it".
